File: backend/src/main/java/com/aimong/backend/global/response/ApiResponse.java
package com.aimong.backend.global.response;

import com.aimong.backend.global.exception.ErrorCode;
import com.fasterxml.jackson.annotation.JsonInclude;

@JsonInclude(JsonInclude.Include.NON_NULL)
public record ApiResponse<T>(
        boolean success,
        T data,
        ErrorBody error,
        String requestId
) {

    public static <T> ApiResponse<T> success(T data) {
        return new ApiResponse<>(true, data, null, RequestIdUtils.getOrCreate());
    }

    public static <T> ApiResponse<T> fail(ErrorCode errorCode) {
        return fail(errorCode, errorCode.getMessage());
    }

    public static <T> ApiResponse<T> fail(ErrorCode errorCode, String message) {
        return new ApiResponse<>(
                false,
                null,
                new ErrorBody(errorCode.getCode(), message),
                RequestIdUtils.getOrCreate()
        );
    }

    public record ErrorBody(
            String code,
            String message
    ) {
    }
}

File: android/app/src/main/java/com/kduniv/aimong/core/network/ApiResponse.kt
package com.kduniv.aimong.core.network

// TODO: sealed class — 네트워크 응답 래퍼
sealed class ApiResponse<out T> {
    data class Success<T>(val data: T) : ApiResponse<T>()
    data class Error(val code: Int, val message: String) : ApiResponse<Nothing>()
    object Loading : ApiResponse<Nothing>()
}
